Doc with explanation of different status codes and error conditions (#37945)
Azure/azure-sdk-for-java: sdk/cosmos/azure-cosmos/docs/StatusCodes.md

# Cosmos DB Java SDK - Status Codes

## Overview

> Below is a summary of the different status code / sub status code combinations that developers can experience in their application when using the Cosmos DB Java SDK. The column "Always transient" indicates whether this error condition is to be expected to always be transient (and auto-mitigated within seconds or at least few minutes). For error conditions that are not always transient the description contains additional context on what to watch out for if this error is seen for longer period of times. For errors that are supposed to be "always transient" it will be necessary to file a support ticket if the error condition exists for an extended period of time. The "Additional Info" column will also point to known issues if available and possible mitigations.

Reviewer: > can experience in their application

We are including combinations that are handled by the SDK internally and do not surface to the customer though. Should we remove those cases from the list? Or convert them to the actual status/substatus that the app level will receive (like the case of 410/21007)?

## Status codes

| Status code | Substatus code | Expecte to be transient | Additional info |
| -----------------:|-----------------------:|:-----------------------:|:----------------------|
|200|0|No| `OK`|
|201|0|No| `Created` - returned for createItem or upsertItem when a new document was created|

> |204|0|No| `No Content` - returned when no payload would ever be returned - like for delete operations|

Author: Added in next iteration

|304|0|No| `Not Modified` - will be returned for `ReadFeed` and `ChangeFeed` operations to indicate that there are no more changes|

> |400|\*|No| `Bad Request` - indicates that the client violated some protocol constraint. See [Bad Request TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-bad-request) for more details.|

Reviewer 2: For larger payload will it be 400 or 413?

> |400|1001|| `Bad Request/Partition key mismatch` - indicates that the PartitionKey defined in the point operation does not match the partition key value being extracted in the service form the document's payload based on the `PartitionKeyDefinition` of the container. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-bad-request) for more details.|

Reviewer: https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-bad-request

> |400|1001|| `Bad Request/Partition key mismatch` - indicates that the PartitionKey defined in the point operation does not match the partition key value being extracted in the service form the document's payload based on the `PartitionKeyDefinition` of the container. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-bad-request) for more details.|

Author: Yes - still calling out the known sub status codes because that was an explicit ask - having them in one place. But the link is added in both cases.

> |400|1001|No| `Bad Request/Partition key mismatch` - indicates that the PartitionKey defined in the point operation does not match the partition key value being extracted in the service form the document's payload based on the `PartitionKeyDefinition` of the container. See [Bad Request TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-bad-request) for more details.|

Reviewer 2: 400/1004: CrossPartitionQueryNotServable from gateway (esp with older SDK's)

Author: Thanks - added in next iteration

> |401|0|No|`Unauthorized` - indicates that the client used invalid credentials. This could mean an invalid key when using `MasterKey`-based authentication, it could mean there is a time-synchronization issue or when using AAD that the AAD credentials are not correctly set-up. See [Unauthorized TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-unauthorized) for more details.|

Reviewer 2: Also worth calling out key-rotation impact that new keys will take upto 10M and might fail intermittently.

Author: Added context around this in the next iteration

|403|\*||`Forbidden` - indicates that the service rejected the request due to missing permissions. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-forbidden)|
|403|3|x|`Forbidden/WriteForbidden` - indicates that the client attempted a write operation against a read-only region in a single write region set-up.|
|403|1008|x|`Forbidden/AccountNotFound` - indicates that the client attempted a read or write operation against a replica that did not have information about the database account.|

> |403|5300||`Forbidden/AADForMetadata` - indicates that the client attempted a metadata operation (like creating, deleting or modifying a container/database) when using AAD authentication. This is not possible via the Data plane SDK. To execute control plane operatiosn with AAD authentication, please use the management SDK. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-forbidden#partition-key-exceeding-storage) and also the [Service quota docuemntation](https://learn.microsoft.com/azure/cosmos-db/concepts-limits#provisioned-throughput) for more details|

Author: Already referenced

> |403|1014||`Forbidden/LogicalPartitionExceedsStorage` - indicates that the data size of a logical partition exceeds the service quota (currently 20 GB). See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-forbidden#non-data-operations-are-not-allowed) for more details|

Reviewer: Included in https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-forbidden already

Author: Already referenced

|404|0||`Not found` - Indicates that the resource the client tried to read does not exist (on the replica being contacted). Depending on the consistency level used this could be a transient error condition - but when using less than strong consistency one that the application needs to be able to handle gracefully. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-not-found) for more details.|

> |404|1002|(x)|`Not Found/Read session no available` - Indicates that a client uses session consitency and reached a replica that has a replication lag and has not caught-up to the requested session token. In many cases this error condition will be transient. But there are certain situation in which it could persist for longer period of times - either a wrong session token is being provided in the application or in a Multi-Write region set-up operations are regulary directed to different regions|

Reviewer 3: ```suggestion
|404|1002|(x)|`Not Found/Read session no available` - Indicates that a client uses session consistency and reached a replica that has a replication lag and has not caught-up to the requested session token. In many cases this error condition will be transient. But there are certain situation in which it could persist for longer period of times - either a wrong session token is being provided in the application or in a Multi-Write region set-up operations are regulary directed to different regions|
```

|404|1003|(x)|`Not Found/Owner resource does not exist` - Indicates that a client attempted to process an operation on a resource whose parent does not exist. For example an attempt to do a point operation on a document when the container does not exist (yet). Can be transient when attempting document operations immediately after creating a container etc. - but when not transient usually means a bug in your application.|

> |404|1024|x|`Not Found/Incorrect Container resource id` - Indicates that a client attempted to use a container that has recently been deleted and rectreated. So, the cached container id in the client is stale - and identifies the previosuly deleted container. The SDK will trigger retries - in general applications need to be able to tolerate that conatiner deletion and immediate recreation will take up-to a few seconds/minutes to be replicated across all regions.|

Reviewer 4: SDK never retries for 404 I think 

for recreate scenario, should the error code be 410/1000?

Author: both are possible - added 1000 as well in next iteration

> |404|1024|x|`Not Found/Incorrect Container resource id` - Indicates that a client attempted to use a container that has recently been deleted and rectreated. So, the cached container id in the client is stale - and identifies the previosuly deleted container. The SDK will trigger retries - in general applications need to be able to tolerate that conatiner deletion and immediate recreation will take up-to a few seconds/minutes to be replicated across all regions.|

Reviewer: What's the difference between `(x)` and `x` in the third column?

Author: Good question - I intended it to mean "usually yes" - but I will refactor and use real words :-)

|408|\*|x|`Request timeout` - Indicates a tiemout for an attempted operation. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-java-sdk-request-timeout) for more details.|

> |408|20008|(x)|`Request timeout/End-to-end timeout exceeded` - Indicates that the application defined end-to-end timeout was exceeded when processing an operation. This will usually be a transient error condition - exceptions are when the application defines unrealistis end-to-end timeouts - for example when executing a query that could very well take a few seconds because it is relatively inefficient or when the end-to-end timeout is lower than the to-be-expected network transit time between the application's location and the Cosmso DB sevice endpoint.|

Reviewer 3: ```suggestion
|408|20008|(x)|`Request timeout/End-to-end timeout exceeded` - Indicates that the application defined end-to-end timeout was exceeded when processing an operation. This will usually be a transient error condition - exceptions are when the application defines unrealistic end-to-end timeouts - for example when executing a query that could very well take a few seconds because it is relatively inefficient or when the end-to-end timeout is lower than the to-be-expected network transit time between the application's location and the Cosmos DB service endpoint.|
```

Author: Fixed in next iteration

|408|20901|No|`Request timeout/Negative End-to-end timeout provided` - Indicates that the application defined a negative end-to-end timeout. This indicates a bug in your application.|

> |409|0|No|`Conflict` - Indicates that the attempt to insert (or upsert) a new document cannot be processed because another document with the same identity (partition key value + value of `id` property) exists or a unqiue key constraint would be violated.|

Reviewer 2: For unique key conflict, will that show-up as 409 with special sub-status code?

Author: No - unfortunately not. Parsing error is the only option - at least until few weeks ago

|410|\*|x|`Gone` - indicates transient error conditions that could happen while replica get moved to a different node or partitions get split/merged. The SDK will retry these error conditions and usually mitigate them without even surfacing them to the application. If these erros get surfaced to the application as `CosmosException` with status code `410` or `503` thses errors should always be transient.|
|410|21010|x|`Service timeout` - Indicates that an operation has been timed out at the service.  See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-request-timeout) for more details. This error will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|

> |410|21001|x|`Name cache is stale` - Indicates that a container was deleted and recreated - and the client's cache still has the old container metadata. This error indicates that the client even after refreshing the cache got the container metadata of the "old" container. Usually it indicates that the replication of the new container metadata across all regions took longer than usual. This error should always be transient. This error will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|

Reviewer 4: they should never see 410/21001 -> we only use this substrates code when bubbling up as a 503. So I wonder we should put into the 503/21001 section

Author: Thanks fixed in next iteration

|410|21003|x|`Completing split` - Indicates that a partition split or merge is pending and commiting the split takes longer than expected. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|
|410|21004|x|`Completing migration` - Indicates that a partition migration due to load-balancing is pending and takes longer than expected. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|
|410|21006|x|`Global strong write barrier not met` - Indicates that synchronous replication of a write operation in a multi-region account with strong consistency did not complete. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|

> |410|21007|x|`Read quorum not met` - Indicates that no read quorum could be achieved when using strong or bounded staleness consistency. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|

Reviewer: Should these be listed as 503 then if what the user will get is a 503? Otherwise one might interpret as a consumer that I need to handle 410/21007

Author: Changed

Author: Actually I think the wording that it will be mapped to 503 when surfacing to app is sufficient

> |410|21007|x|`Read quorum not met` - Indicates that no read quorum could be achieved when using strong or bounded staleness consistency. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|

Reviewer 5: This line is repeated.

Author: Fixed in next iteration

> |412|0||`Precondition failed` - The document has been modified since the application read it (and retrieved the etag that was used as pre-codnition for the write operation). This is the typical optimistic concurrency signal - and needs to be gracefully handled in your application. The usual patterns is to re-read the document, apply teh same changes and retry the write with the updated etag.|

Reviewer 4: typo:  apply `teh` same changes ->  apply `the` same changes

> |412|0||`Precondition failed` - The document has been modified since the application read it (and retrieved the etag that was used as pre-codnition for the write operation). This is the typical optimistic concurrency signal - and needs to be gracefully handled in your application. The usual patterns is to re-read the document, apply teh same changes and retry the write with the updated etag.|

Reviewer 4: ```suggestion
|412|0||`Precondition failed` - The document has been modified since the application read it (and retrieved the etag that was used as pre-codnition for the write operation). This is the typical optimistic concurrency signal - and needs to be gracefully handled in your application. The usual patterns is to re-read the document, apply the same changes and retry the write with the updated etag.|
```

Reviewer: Add a link to https://aka.ms/CosmosDB/sql/errors/precondition-failed

Author: Fixed in next iteration

> |500|0|Unknown|`Internal Server error` - Indicates unexpected and unqualified internal service error.|

Reviewer 2: From service telemetry 500 is a small fraction compated to 503, approx 5%.
Also HTTP spec wise 503 is more specific and say that service is unavailable. How about recommended retry on 500 also?

Author: Agreed - but that should go into the "building resilient apps" doc - don't want to clutter any instructions when to do app-level retries here. Also called this out explicitly in the overview/out of scope section.

> |429|3200||`User throttling` - Indicates that the operations being processed by your Cosmso DB account exceed the provisioned throughptu RU/s. Mitigation can be done by either scaling-up - or improving the efficiency especially of queries to reduce the RU/s consumption. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-request-rate-too-large) for more details.|

Reviewer 4: ```suggestion
|429|3200||`User throttling` - Indicates that the operations being processed by your Cosmso DB account exceed the provisioned throughput RU/s. Mitigation can be done by either scaling-up - or improving the efficiency especially of queries to reduce the RU/s consumption. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-request-rate-too-large) for more details.|
```

Author: Fixed in next iteration

> |429|3201||`Metadata throttling` - Indicates that metadata operations are being throttled. Increasing provisioned throughput (RU/s) won't help - this usually indicates a bug in your application where meatdata calls are triggered extensively or you are not using a singleton pattern for `CosmosClinet`/`CosmosAsyncClient`. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-request-rate-too-large) for more details.|

Reviewer 4: ```suggestion
|429|3201||`Metadata throttling` - Indicates that metadata operations are being throttled. Increasing provisioned throughput (RU/s) won't help - this usually indicates a bug in your application where metadata calls are triggered extensively or you are not using a singleton pattern for `CosmosClinet`/`CosmosAsyncClient`. See [TSG](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-request-rate-too-large) for more details.|
```

Author: Fixed in next iteration

> |429|< 3200|x||`SLA violating throttling` - Indicates service-side throttling that will count against the service's SLA. These errors should always be transient and will surface as a CosmosException with status code `503` after exceeding SDK-retries| 

Author: Yeah - copy and paste error - fixed in next iteration

|449|0|x||`RetryWith` - Indicates a concurrent attempt to chaneg documents server-side - for example via patch or stored procedure invocation. The `449` status code will be automatically retried by the SDK. This condition should always be transient as long as the application is not excessively doing concurrent changes to documents.|

> |500|0||`Internal Server error` - Indicates unexpected and unqualified internal service error. This condition should always be transient.|

Reviewer 4: 500 may not be always transient, could be due to a bug?

Author: Changed in next iteration

|502|0||`Bad gateway` - Indicated an HTTP proxy you are using is misbehaving. In general HTTP proxies are not recommended for any production workload.|
|503|21005|x|`Serverside 410` - Indicates that a replica returns a 410 - usually during initialization of the replica. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|

> |503|21008|x|`Service unavailable` - Indicates that a replica returned `503` service unavailable. This error should always be transient and will surface as a CosmosException with status code `503` after exceeding SDK-retries.|

Reviewer: Should we treat 503 as a blanket?

Author: Changed in next iteration

> |504|0|Unknown|`Gateway timeout` - Indicated an HTTP proxy you are using timed  out. In general HTTP proxies are not recommended for any production workload.|

Reviewer 2: thoughts on call out that its not part of CosmosDB or related explicitly

Reviewer 2: Same for 502

Author: Done in next iteration

# Cosmos DB Java SDK - Status Codes

## Overview

Below is a list of the different status code / sub status code combinations that admins and/or developers can experience in their application or when looking at diagnostics/metrics when using the Cosmos DB Java SDK.

> Several of these error conditions would never surface to the application, because the SDK has built-in retry-logic to recover. But these status codes could show up in diagnostics and/or micrometer metrics - and there has been the ask to provide some context around them - like:

Author: Fixed

- what do these error conditions mean - even when they are automatically handled by the SDK?
- how long should a certain error condition happen before self-recovery happens?

> The column "Expected to be transient" indicates whether this error condition is to be expected to always be seen transiently  (and auto-mitigated within seconds or at least few minutes should happen in the SDK and/or the service). For error conditions that might not go away automatically the description contains additional context on what to watch out for if this error is seen for longer period of times. For errors that are supposed to be "transient" it will be necessary to file a support ticket, if the error condition exists for an extended period of time. The "Additional Info" column will also point to known issues if available and possible mitigations.

Author: Changed to...

```
The column "Expected to be transient" indicates whether this error condition is to be expected to always be seen for a short period of time (and auto-mitigated within seconds or at least few minutes should happen in the SDK and/or the service).
```

## Out of scope

> This document is intentionally not goind into details on how resilient applications should react  to certain error conditions (and whether/how retries are recommended). There is prescriptive guidance for developers around this located here: [Designing resilient applications with Azure Cosmos DB SDKs](https://learn.microsoft.com/azure/cosmos-db/nosql/conceptual-resilient-sdk-applications)

Author: Fixed

|401|0|No|`Unauthorized` - indicates that the client used invalid credentials. The most frequent scenario when this is happening, is when customers rotate the currently used key. Key rotation needs to be replicated across regions, which can take up-to a few minuts. During this time a `401 Unauthroized` would be used when the client is using the old or new key while the replication is still happening. The best way to do key rotation is to rotate the key only after it is not used by applications anymore - that is why a primary and secondary key exists for both writable and read-only keys. More details can be found here - [key rotation best practices](https://learn.microsoft.com/azure/cosmos-db/secure-access-to-data?tabs=using-primary-key#key-rotation). In addition this could also mean an invalid key when using `MasterKey`-based authentication, it could mean there is a time-synchronization issue or when using AAD that the AAD credentials are not correctly set-up. See [Unauthorized TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-unauthorized) for more details.|

> |403|\*|No|`Forbidden` - indicates that the service rejected the request due to missing permissions. See [Forbidden TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-forbidden)|

Author: That is why is spelled the acronym out explicitly?

|404|1002|In most cases|`Not Found/Read session no available` - Indicates that a client uses session consistency and reached a replica that has a replication lag and has not caught-up to the requested session token. In many cases this error condition will be transient. But there are certain situation in which it could persist for longer period of times - either a wrong session token is being provided in the application or in a Multi-Write region set-up operations are regulary directed to different regions|
|404|1003|Yes (up to few minutes)|`Not Found/Owner resource does not exist` - Indicates that a client attempted to process an operation on a resource whose parent does not exist. For example an attempt to do a point operation on a document when the container does not exist (yet). Can be transient when attempting document operations immediately after creating a container etc. - but when not transient usually means a bug in your application.|

> |404|1000 or 1024|x|`Not Found/Incorrect Container resource id` - Indicates that a client attempted to use a container that has recently been deleted and rectreated. So, the cached container id in the client is stale - and identifies the previosuly deleted container. The SDK will trigger retries - in general applications need to be able to tolerate that conatiner deletion and immediate recreation will take up-to a few seconds/minutes to be replicated across all regions.|

Reviewer 6: typos: 
change "rectreated" to "recreated"
change "conatiner" to "container"

Author: Fixed

> |404|1000 or 1024|x|`Not Found/Incorrect Container resource id` - Indicates that a client attempted to use a container that has recently been deleted and recreated. So, the cached container id in the client is stale - and identifies the previosuly deleted container. The SDK will trigger retries - in general applications need to be able to tolerate that container deletion and immediate recreation will take up-to a few seconds/minutes to be replicated across all regions.|

Reviewer 4: it will be 410/1000 I think 

also curious when/forWhichOperation we are going to see 404/1024? (I have not seen one)

Author: Ok, I split them. 

404/1024 can happen in Gateway mode

From ClientRetryPolicy

```
        // This is for gateway mode, collection recreate scenario is not handled there
        if (clientException != null &&
            Exceptions.isStatusCode(clientException, HttpConstants.StatusCodes.BADREQUEST) &&
            Exceptions.isSubStatusCode(clientException, HttpConstants.SubStatusCodes.INCORRECT_CONTAINER_RID_SUB_STATUS)) {
            return this.shouldRetryOnStaleContainer();
        }
```

|408|20008|Yes, unless unrealistic e2e timeout is used|`Request timeout/End-to-end timeout exceeded` - Indicates that the application defined end-to-end timeout was exceeded when processing an operation. This will usually be a transient error condition - exceptions are when the application defines unrealistic end-to-end timeouts - for example when executing a query that could very well take a few seconds because it is relatively inefficient or when the end-to-end timeout is lower than the to-be-expected network transit time between the application's location and the Cosmos DB service endpoint.|
|408|20901|No|`Request timeout/Negative End-to-end timeout provided` - Indicates that the application defined a negative end-to-end timeout. This indicates a bug in your application.|
|409|0|No|`Conflict` - Indicates that the attempt to insert (or upsert) a new document cannot be processed because another document with the same identity (partition key value + value of `id` property) exists or a unqiue key constraint would be violated.|

> |410|\*|Yes|`Gone` - indicates transient error conditions that could happen while replica get moved to a different node or partitions get split/merged. The SDK will retry these error conditions and usually mitigate them without even surfacing them to the application. If these erros get surfaced to the application as `CosmosException` with status code `410` or `503` thses errors should always be transient.|

Reviewer 6: typo: "erros" to "errors"

Author: Fixed

|503|21002|Yes|`Partition key range gone` - Indicates that a partition split or merge happened and the client even after several retries was not able to get the metadata for the new partition. This error indicates a delay of replication of partition key range metadata and should always be transient.|
|503|21003|Yes|`Completing split` - Indicates that a partition split or merge is pending and commiting the split takes longer than expected. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|
|503|21004|Yes|`Completing migration` - Indicates that a partition migration due to load-balancing is pending and takes longer than expected. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|

> |503|21005|Yes|`Serverside 410` - Indicates that a replica returns a 410 - usually during initialization of the replica. This error should always be transient and will be mapped to a CosmosException with status code `503` when surfacing it to the application after exceeding SDK-retries.|

Reviewer 4: 410/21005 will be seen in diagnostics as well

## Out of scope
This document is intentionally not going into details on how resilient applications should react  to certain error conditions (and whether/how retries are recommended). There is prescriptive guidance for developers around this located here: [Designing resilient applications with Azure Cosmos DB SDKs](https://learn.microsoft.com/azure/cosmos-db/nosql/conceptual-resilient-sdk-applications)

## Status codes

| Status code | Substatus code | Expected to be transient | Additional info |
| -----------------:|-----------------------:|:-----------------------:|:----------------------|
|200|0|No| `OK`|
|201|0|No| `Created` - returned for createItem or upsertItem when a new document was created|
|204|0|No| `No Content` - returned when no payload would ever be returned - like for delete operations|
|207|0|No| `Multi-Status` - returned for bulk operations when some of the item operations have failed and others succeeded. The API allows checking status codes of item operations.|

> |304|0|No| `Not Modified` - will be returned for `ReadFeed` and `ChangeFeed` operations to indicate that there are no more changes|

Reviewer: Technically this is only for Change Feed. ReadFeed appears in our metrics for Change Feed requests, but its not really a ReadFeed.

|400|\*|No| `Bad Request` - indicates that the client violated some protocol constraint. See [Bad Request TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-bad-request) for more details.|
|400|1001|No| `Bad Request/Partition key mismatch` - indicates that the PartitionKey defined in the point operation does not match the partition key value being extracted in the service form the document's payload based on the `PartitionKeyDefinition` of the container. See [Bad Request TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-bad-request) for more details.|

> |400|1004|No| `Bad Request/CrossPartitionQueryNotServable` - indicates that the client attempted to execute a cross-partition query, which cannot be processed with the current SDK version. Usually this means that the query uses a query construct, which is not yet supported in the SDK version being used. Upgrading the SDK might help to address the problem. See [Bad Request TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-bad-request) for more details.|

Reviewer: nit: Does this currently happen in the SDK? With the existence of the QueryPlan request. I remember this used to happen on older SDKs.

Author: Yes - I has the same thought when Kiran raised it yesterday. But given that we probably want to generalize this anyway across SDKs I added it, what is the harm when the error never happens ?

|403|\*|No|`Forbidden` - indicates that the service rejected the request due to missing permissions. See [Forbidden TSG - trouble-shooting guide](https://learn.microsoft.com/azure/cosmos-db/nosql/troubleshoot-forbidden)|

> |403|3|Yes (up to few minutes)|`Forbidden/WriteForbidden` - indicates that the client attempted a write operation against a read-only region in a single write region set-up.|

Reviewer: nit: `(up to few minutes)` This might be depending on what the user does, if they keep adding or removing region or failing over, 403/3 and 403/1008 could extend over time

Author: True - but I think it is safe to assume that this won't cause too may CRIs :-) - So leaving as is.

|408|20008|Yes, unless unrealistic e2e timeout is used|`Request timeout/End-to-end timeout exceeded` - Indicates that the application defined end-to-end timeout was exceeded when processing an operation. This will usually be a transient error condition - exceptions are when the application defines unrealistic end-to-end timeouts - for example when executing a query that could very well take a few seconds because it is relatively inefficient or when the end-to-end timeout is lower than the to-be-expected network transit time between the application's location and the Cosmos DB service endpoint.|
|408|20901|No|`Request timeout/Negative End-to-end timeout provided` - Indicates that the application defined a negative end-to-end timeout. This indicates a bug in your application.|
|409|0|No|`Conflict` - Indicates that the attempt to insert (or upsert) a new document cannot be processed because another document with the same identity (partition key value + value of `id` property) exists or a unqiue key constraint would be violated.|

> |410|\*|Yes|`Gone` - indicates transient error conditions that could happen while replica get moved to a different node or partitions get split/merged. The SDK will retry these error conditions and usually mitigate them without even surfacing them to the application. If these errors get surfaced to the application as `CosmosException` with status code `410` or `503` thses errors should always be transient.|

Reviewer: ```suggestion
|410|\*|Yes|`Gone` - indicates transient error conditions that could happen while replica get moved to a different node or partitions get split/merged. The SDK will retry these error conditions and usually mitigate them without even surfacing them to the application. If these errors get surfaced to the application as `CosmosException` with status code `410` or `503` these errors should always be transient.|
```

Author: Fixed
